lean4/src/putnam_1984_a6.lean

```diff
@@ -1,18 +1,18 @@
 import Mathlib
 open BigOperators
 
-open Topology Filter Nat
+open Topology Filter Function Nat
 
--- Note: uses (ℕ → ℕ) instead of (Fin k → ℕ)
-abbrev putnam_1984_a6_solution : Prop × ℕ := sorry
--- (True, 4)
+-- Note: uses `0` to indicate not periodic.
+abbrev putnam_1984_a6_solution : ℕ := sorry
+-- 4
 theorem putnam_1984_a6
-(f : ℕ → ℕ)
-(kadistinct : ℕ → (ℕ → ℕ) → Prop)
-(hkadistinct : kadistinct = fun k : ℕ => fun a : ℕ → ℕ => (k ≥ 1 ∧ ∀ i j : Fin k, i ≠ j → a i ≠ a j))
-(gpeq : (ℕ → ℕ) → ℕ → Prop)
-(hgpeq : gpeq = fun g : ℕ → ℕ => fun p : ℕ => (p > 0 ∧ ∀ s ≥ 1, g s = g (s + p)))
-(hf : ∀ n > 0, f n = if (n % 10) ≠ 0 then n % 10 else f (n / 10))
-: ∃ g : ℕ → ℕ, (∀ (k : ℕ) (a : ℕ → ℕ), kadistinct k a → g (∑ i : Fin k, a i) = f (∑ i : Fin k, 5 ^ (a i))) ∧
-(if putnam_1984_a6_solution.1 = True then (gpeq g putnam_1984_a6_solution.2 ∧ ∀ p : ℕ, gpeq g p → p ≥ putnam_1984_a6_solution.2) else (¬∃ p : ℕ, gpeq g p)) :=
-sorry
+    (f : ℕ → ℕ)
+    (hf : ∀ n, some (f n) = (Nat.digits 10 (n !)).find? (fun d ↦ d ≠ 0))
+    (P : (ℕ → ℕ) → ℕ → Prop)
+    (P_def : ∀ g p, P g p ↔
+      if p = 0 then (∀ q > 0, ¬ Periodic g q) else (Periodic g p ∧ ∀ q > 0, Periodic g q → p ≤ q)) :
```

Comment:
Might this not be a case to use `IsLeast`?

Reply:
Good idea; done.